hpx/runtime/actions/continuation.hpp

```diff
@@ -925,6 +939,14 @@ namespace hpx { namespace actions
     template <typename Arg0>
     void continuation::trigger(Arg0 && arg0)
     {
+        // The dynamic cast decays the argument type to avoid the assert firing
+        // for cases when Arg0 is a const&. This does not make the code invalid
+        // as trigger_value (which is a virtual function) takes its argument
+        // by && anyways.
+        HPX_ASSERT(0 != dynamic_cast<
+                typed_continuation<typename util::decay<Arg0>::type> *
+            >(this));
+
```

Comment:
It's difficult to explain. Let's assume `Arg0` is deduced as `id_type const&`. In this case, the `dynamic_cast<Arg0>(this)` would fail as `this` is not of type `typed_continuation<id_type const&>` but of type `typed_continuation<id_type>`. Hence the decayed argument.

At the same time, if we statically cast `this` to `typed_continuation<id_type>`, then its member function `trigger_value` effectively would have the signature `trigger_value(id_type&&)` causing compilation problems (remember, `trigger_value` is a virtual function and we have no way to use a forwarding reference here - which would solve the issue).

One solution would be to have another virtual function `trigger_value(Result const&)` which would be used by the compiler in this case. Adding this however causes issues whenever `Arg0` represents a move-only type (we can't 'disable' the virtual function from the type for this case).

This code is fishy as it does a `static_cast<typed_continuation<id_type const&>>(this)` where `this` is actually a `typed_continuation<id_type>`. This however should not be a problem as for both types the effectively generated class is 100% identical in layout and member function signatures.

Please let me know if you have any other suggestion on how to resolve the problem.